get_mvn.go

```diff
@@ -0,0 +1,172 @@
+	}
+
+	version := q.Get("version")
+	if version == "" {
+		return "", fmt.Errorf("query parameter 'version' is required.")
+	}
+
+	classifier := q.Get("classifier")
+
+	artType := q.Get("type")
+	if artType == "" {
+		artType = "jar"
+	}
+
+	filename := artifactId + "-" + version
+	if classifier != "" {
+		filename += "-" + classifier
+	}
+	filename += "." + artType
+
+	return filename, nil
+}
+
+func (g *MvnGetter) Get(dstDir string, u *url.URL) error {
+	filename, err := g.GetFilename(u)
+	if err != nil {
+		return err
+	}
+
+	dstFile := filepath.Join(dstDir, filename)
+	return g.GetFile(dstFile, u)
+}
+
+// Get the artifact from remote maven repo.
+//
+// If the version is a snapshot version, it will get the latest snapshot artifact.
+// Query parameters:
+//   - groupId: the group id
+//   - artifactId: the artifact id
+//   - version: the artifact version
+//   - type: the artifact type, default as 'jar'
+// example url: mvn::http://username@host/mavan/repo/path?groupId=org.example&artifactId=test&version=1.0.0-SNAPSHOT
+func (g *MvnGetter) GetFile(dst string, u *url.URL) error {
+	q := u.Query()
+	groupId := q.Get("groupId")
+	if groupId == "" {
+		return fmt.Errorf("query parameter 'groupId' is required.")
+	}
+	artifactId := q.Get("artifactId")
+	if artifactId == "" {
+		return fmt.Errorf("query parameter 'artifactId' is required.")
+	}
+	// the artifact version, Ex., 6.13.1 or 6.13-SNAPSHOT
+	version := q.Get("version")
+	if version == "" {
+		return fmt.Errorf("query parameter 'version' is required.")
+	}
+	classifier := q.Get("classifier")
+	artType := q.Get("type")
+	if artType == "" {
+		artType = "jar"
+	}
+
+	// construct the real url hits the maven repo
+	artifactUrl, err := url.Parse(u.String())
+	if err != nil {
+		return err
+	}
+	artifactUrl.RawQuery = ""
+	artifactUrl.Path = path.Join(artifactUrl.Path, fmt.Sprintf("/%s/%s/%s", strings.Replace(groupId, ".", "/", -1), artifactId, version))
+
+	// the artifact file version.
+	//   when the artifact version is a snapshot version, the artifact file version will be expanded to the latest snapshot version, Ex., '6.13-20171126.202552-6'
+	artifactFileVer := version
+	if strings.HasSuffix(version, "-SNAPSHOT") {
+		// get the latest snapshot
+		snapshotVer, err := g.ParseLastestSnapshotVersion(artifactUrl)
+		if err != nil {
+			return err
+		}
+
+		artifactFileVer = snapshotVer
+	}
+
+	filename := artifactId + "-" + artifactFileVer
+	if classifier != "" {
+		filename += "-" + classifier
+	}
+	filename += "." + artType
+	artifactUrl.Path = path.Join(artifactUrl.Path, filename)
+
+	return g.HttpGet.GetFile(dst, artifactUrl)
+}
+
+// get the latest snapshot version by parsig the maven-metadata.xml from remote maven repo.
+//   - artifactVerUrl the url to the artifact version, Ex., 'https://repo1.maven.org/maven2/org/testng/testng/6.13.1/'
+func (g *MvnGetter) ParseLastestSnapshotVersion(artifactVerUrl *url.URL) (string, error) {
```

Comment:
@schmichael , this method `ParseLastestSnapshotVersion ` is used by the test code, so that's why it's exposed.

Reply:
Since the test is in the same package this does not need to be exported.